src/storage.rs

```diff
@@ -1,9 +1,58 @@
 //! Storage backend implementations.
 
+use core::marker::PhantomData;
+use fuel_storage::Mappable;
+use fuel_tx::Contract;
+use fuel_types::{AssetId, Bytes32, ContractId, Salt, Word};
+
 mod interpreter;
 mod memory;
 mod predicate;
 
 pub use interpreter::InterpreterStorage;
 pub use memory::MemoryStorage;
 pub use predicate::PredicateStorage;
+
+/// The storage table for contract's raw byte code.
+pub struct ContractsRawCode;
+
+impl Mappable for ContractsRawCode {
+    type Key = ContractId;
+    type SetValue = [u8];
+    type GetValue = Contract;
+}
+
+/// The storage table for contract's additional information as salt, root hash, etc.
+pub struct ContractsInfo;
+
+impl Mappable for ContractsInfo {
+    type Key = ContractId;
+    /// `Salt` - is the salt used during creation of the contract for uniques.
+    /// `Byte32` - is the root hash of the contract's code.
+    type SetValue = (Salt, Bytes32);
+    type GetValue = Self::SetValue;
+}
+
+/// The storage table for contract's assets balances.
+///
+/// Lifetime is for optimization to avoid `clone`.
+pub struct ContractsAssets<'a>(PhantomData<&'a ()>);
```

Comment:
Can't this be a tuple of generics that implements either `AsRef<ContractId>` or `Borrow<ContractId>`? Introducing lifetime requirements to the signature might increase the complexity of the implementation and I think it should be avoided whenever possible so we don't confuse the compiler about ownership requirements

Reply:
It can be, but in this case, the definition will look like this:

<img width="534" alt="image" src="https://user-images.githubusercontent.com/18346821/189171717-cedeaff6-6d81-412e-8e83-23b82834f161.png">

And you need to propagate those generics everywhere. Instead of `StorageMutate<ContractsAssets<'_>>`, you need to write `StorageMutate<ContractsAssets<C, A>>`, and `C` and `A` should be generics with bounds.

Plus compiler will not force you to use reference, and sometimes you can use `Copy` in places where you don't want to do that. 

But opposite it, you can have cases where you want to pass by reference and pass by value. In these cases, generic with `Borrow` will be better.

The usage of an inferred lifetime(`'_`) almost in all cases will work for you(I guess in all, but who knows). The same way references in the methods work. So you will not have errors related to it. Also, you can write `self.storage::<ContractsAsset>()`, and it works.

So, if we want to support pass by reference and value, I can use generics(but it means propagation of them). If not, the usage of a lifetime `'_` will not harm us=)

Reply:
I'm fine going with the reference, but I'd rather go with the generic+borrow because this is more canonical to the language & more flexible (we will be able to use anything that can be a contract id; from Cow to references).

You're right, the `'_` should cover most of the cases